rename Invokers features from invoke to command

keithamus opened the pull request:
#### 99b3a16505f8e33c592e4b2981d150485cb2f9fd
<pre>
rename Invokers features from invoke to command
<a href="https://bugs.webkit.org/show_bug.cgi?id=276616">https://bugs.webkit.org/show_bug.cgi?id=276616</a>

Reviewed by Anne van Kesteren.

This renames the attributes invoketarget-&gt;commandfor and invokeaction-&gt;command,
as well as renaming the InvokeEvent to CommandEvent, and it&apos;s action property
to command. It also drops the `auto` action, requiring explicit
`command` attribute for each action.

Lastly, this also adds the logic preventing invokers from being triggered within
a form without an explicit type=button attribute.

These changes are based on a set of discussions around the proposals
naming and semantics, which starts here: <a href="https://github.com/whatwg/html/issues/9625#issuecomment-2115718679">https://github.com/whatwg/html/issues/9625#issuecomment-2115718679</a>
and concludes here; <a href="https://github.com/whatwg/html/issues/9625#issuecomment-2209335876">https://github.com/whatwg/html/issues/9625#issuecomment-2209335876</a>

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/idlharness.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invokeelement-interface.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invokeevent-dispatch-shadow.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invokeevent-interface.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-button-event-dispatch.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-dialog-behavior.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-popover-behavior.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-popover-invalid-behavior.tentative-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-dialog-behavior.tentative-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-dialog-behavior.tentative-expected.txt.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/SmartPointerExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/CommandEvent.cpp: Renamed from Source/WebCore/dom/InvokeEvent.cpp.
(WebCore::CommandEvent::CommandEvent):
(WebCore::CommandEvent::create):
(WebCore::CommandEvent::createForBindings):
(WebCore::CommandEvent::isCommandEvent const):
(WebCore::CommandEvent::invoker const):
* Source/WebCore/dom/CommandEvent.h: Renamed from Source/WebCore/dom/InvokeEvent.h.
* Source/WebCore/dom/CommandEvent.idl: Renamed from Source/WebCore/dom/InvokeEvent.idl.
* Source/WebCore/dom/DeviceOrientationAndMotionAccessController.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::handlePopoverLightDismiss):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isElementReflectionAttribute):
* Source/WebCore/dom/Element.h:
(WebCore::Element::isValidCommandType):
(WebCore::Element::handleCommandInternal):
(WebCore::Element::isValidInvokeAction): Deleted.
(WebCore::Element::handleInvokeInternal): Deleted.
* Source/WebCore/dom/Event.h:
(WebCore::Event::isCommandEvent const):
(WebCore::Event::isInputEvent const):
(WebCore::Event::isInvokeEvent const): Deleted.
* Source/WebCore/dom/EventInterfaces.in:
* Source/WebCore/dom/EventNames.json:
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::defaultEventHandler):
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::isValidCommandType):
(WebCore::HTMLDialogElement::handleCommandInternal):
(WebCore::HTMLDialogElement::isValidInvokeAction): Deleted.
(WebCore::HTMLDialogElement::handleInvokeInternal): Deleted.
* Source/WebCore/html/HTMLDialogElement.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::isValidCommandType):
(WebCore::HTMLElement::handleCommandInternal):
(WebCore::HTMLElement::isValidInvokeAction): Deleted.
(WebCore::HTMLElement::handleInvokeInternal): Deleted.
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::commandForElement const):
(WebCore::HTMLFormControlElement::commandType const):
(WebCore::HTMLFormControlElement::handleCommand):
(WebCore::HTMLFormControlElement::invokeTargetElement const): Deleted.
(WebCore::HTMLFormControlElement::invokeAction const): Deleted.
(WebCore::HTMLFormControlElement::handleInvokeAction): Deleted.
* Source/WebCore/html/HTMLFormControlElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::defaultEventHandler):
* Source/WebCore/html/InvokerElement.idl:
* Source/WebCore/page/FragmentDirective.h:

Canonical link: <a href="https://commits.webkit.org/281345@main">https://commits.webkit.org/281345@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e04ceeb3446a6d9ad0bcd9ca6937743a035bcd1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59596 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38941 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12120 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63511 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10119 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61725 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46594 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10272 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48354 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7085 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61626 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36346 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51586 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29192 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33049 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8838 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9043 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/52689 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54996 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9116 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65243 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/58841 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3524 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9006 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55695 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3535 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51579 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55822 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13206 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2925 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/80598 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34755 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13984 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35838 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36924 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35583 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/SmartPointerExpectations/UncountedLocalVarsCheckerExpectations:
```diff
@@ -1399,4 +1399,5 @@ xml/XSLStyleSheetLibxslt.cpp
 xml/XSLTProcessor.cpp
 xml/XSLTProcessorLibxslt.cpp
 xml/parser/XMLDocumentParser.cpp
-xml/parser/XMLDocumentParserLibxml2.cpp
```
Please remove this change.

annevk commented on Source/WebCore/dom/EventInterfaces.in:
```diff
@@ -13,10 +13,12 @@ BeforeUnloadEvent
 CookieChangeEvent
 CSSAnimationEvent
 CSSTransitionEvent
+CommandEvent
```
Seems to me you should not include it twice.

annevk commented on Source/WebCore/dom/EventNames.json:
```diff
@@ -318,4 +318,4 @@
     "writeend": { },
     "writestart": { },
     "zoom": { }
-}
```
Please don't change this.

annevk commented on Source/WebCore/html/HTMLButtonElement.cpp:
```diff
@@ -140,7 +140,13 @@ void HTMLButtonElement::defaultEventHandler(Event& event)
+
+            handleCommand();
+
+        } if (protectedForm) {
```
Should this be else if? This doesn't make much sense.

annevk commented on Source/WebCore/dom/Element.h:
```diff
@@ -109,10 +109,9 @@ enum class ShadowRootSerializable : bool { No, Yes };
+enum class CommandType: uint8_t {
     Invalid,
 
-    Auto,
```
Was the corresponding code already removed?

keithamus replied:
The only reference to auto was just below on [L658](https://github.com/WebKit/WebKit/pull/30825/files/1f6eaeae7b9c7fe24771d4039630fec066e79b72#diff-80e7accc2bd275dc38997e95f574e0bf7972e13e89c56d5761d883ff90507ec1L658)

annevk commented on Source/WebCore/html/HTMLFormControlElement.cpp:
```diff
@@ -423,69 +423,69 @@ RefPtr<Element> HTMLFormControlElement::invokeTargetElement() const
 constexpr ASCIILiteral showModalLiteral = "showmodal"_s;
 constexpr ASCIILiteral closeLiteral = "close"_s;
-InvokeAction HTMLFormControlElement::invokeAction() const
+CommandType HTMLFormControlElement::commandType() const
```
Nit: `std::optional<CommandType>` would be slightly cleaner here as opposed to `CommandType::Invalid`. (Okay as follow-up.)

annevk commented on Source/WebCore/html/InvokerElement.idl:
```diff
@@ -25,6 +25,6 @@
 
 [EnabledBySetting=InvokerAttributesEnabled]
 interface mixin InvokerElement {
```
Shouldn't this be renamed as well?

keithamus replied:
Would `InvokerCommandElement` be a suitable name? or would you prefer something like `CommandButtonElement`?

annevk replied:
I guess this part we should first discuss in the WHATWG as we might well be able to flatten it into HTMLButtonElement.

keithamus replied:
Given the conclusion in whatwg, I'll follow up with some of these changes.

annevk commented on Source/WebCore/html/HTMLButtonElement.cpp:
```diff
@@ -140,7 +140,13 @@ void HTMLButtonElement::defaultEventHandler(Event& event)
 
-        if (protectedForm) {
+        if (commandForElement()) {
+            if (m_type != BUTTON && form())
+                return;
```
I forgot what we decided here. This means that `<button type=submit commandfor=i>` no longer submits, even though there's no `command` attribute and even though `type=submit` can't be a command button anyway. Is that really what we want?

annevk replied:
For the `input` element it seems we're doing something completely different. Hmm.

keithamus replied:
From my recollection we had decided that submit buttons with `commandfor` would no-op, to ward folks away from what was likely an author error.

annevk replied:
I guess it's reasonable if that happens even when `type` is not defaulted but set explicitly, as it's still an error.

I guess what I'm less sure on is whether both `command` and `commandfor` have to be set for that or if just `commandfor` suffices. As `<button type=button commandfor=i>` won't do anything either.

keithamus replied:
Chrome logs some of these kinds of author errors to the console as warnings. I'm happy to do the same here but it doesn't seem like a convention in WebKit?

annevk replied:
I think we should do it where it seems useful. It might well be for cases like this.

keithamus replied:
I'll follow up with this.

annevk commented on Source/WebCore/dom/CommandEvent.idl:
```diff
@@ -26,14 +26,14 @@
+] interface CommandEvent : Event {
+    constructor([AtomString] DOMString type, optional CommandEventInit eventInitDict);
 
     readonly attribute Element? invoker;
```
I don't think we should continue calling this "invoker". But I guess that can be a follow-up? Maybe just `button`.

annevk replied:
Or `commandElement` as counterpart to `commandForElement`. I kinda like that as well.

keithamus replied:
Given the conclusion in whatwg, I'll follow up with some of these changes.

annevk commented on Source/WebCore/html/HTMLFormControlElement.cpp:
```diff
@@ -423,69 +423,69 @@ RefPtr<Element> HTMLFormControlElement::invokeTargetElement() const
+void HTMLFormControlElement::handleCommand()
 {
-    RefPtr invokee = invokeTargetElement();
+    RefPtr invokee = commandForElement();
```
`target` would be a clearer variable name here.

keithamus replied:
I'll be sure to include this in the follow up PR.

annevk commented on Source/WebCore/html/HTMLFormControlElement.cpp:
```diff
@@ -423,69 +423,69 @@ RefPtr<Element> HTMLFormControlElement::invokeTargetElement() const
+CommandType HTMLFormControlElement::commandType() const
 {
-    auto action = attributeWithoutSynchronization(HTMLNames::invokeactionAttr);
+    auto action = attributeWithoutSynchronization(HTMLNames::commandAttr);
```
This should probably be named `command`? Although the specification PR still calls some things action?

annevk reviewed (approved):
Once rebased and passing all the bots this seems okay to land. It does seem some further renames and possibly cleanup will be required before this can be enabled more widely.

keithamus committed:
rename Invokers features from invoke to command
https://bugs.webkit.org/show_bug.cgi?id=276616

Reviewed by Anne van Kesteren.

This renames the attributes invoketarget->commandfor and invokeaction->command,
as well as renaming the InvokeEvent to CommandEvent, and it's action property
to command. It also drops the `auto` action, requiring explicit
`command` attribute for each action.

Lastly, this also adds the logic preventing invokers from being triggered within
a form without an explicit type=button attribute.

These changes are based on a set of discussions around the proposals
naming and semantics, which starts here: https://github.com/whatwg/html/issues/9625#issuecomment-2115718679
and concludes here; https://github.com/whatwg/html/issues/9625#issuecomment-2209335876

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/idlharness.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invokeelement-interface.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invokeevent-dispatch-shadow.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invokeevent-interface.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-button-event-dispatch.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-dialog-behavior.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-popover-behavior.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-popover-invalid-behavior.tentative-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-dialog-behavior.tentative-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-dialog-behavior.tentative-expected.txt.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/SmartPointerExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/CommandEvent.cpp: Renamed from Source/WebCore/dom/InvokeEvent.cpp.
(WebCore::CommandEvent::CommandEvent):
(WebCore::CommandEvent::create):
(WebCore::CommandEvent::createForBindings):
(WebCore::CommandEvent::isCommandEvent const):
(WebCore::CommandEvent::invoker const):
* Source/WebCore/dom/CommandEvent.h: Renamed from Source/WebCore/dom/InvokeEvent.h.
* Source/WebCore/dom/CommandEvent.idl: Renamed from Source/WebCore/dom/InvokeEvent.idl.
* Source/WebCore/dom/DeviceOrientationAndMotionAccessController.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::handlePopoverLightDismiss):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isElementReflectionAttribute):
* Source/WebCore/dom/Element.h:
(WebCore::Element::isValidCommandType):
(WebCore::Element::handleCommandInternal):
(WebCore::Element::isValidInvokeAction): Deleted.
(WebCore::Element::handleInvokeInternal): Deleted.
* Source/WebCore/dom/Event.h:
(WebCore::Event::isCommandEvent const):
(WebCore::Event::isInputEvent const):
(WebCore::Event::isInvokeEvent const): Deleted.
* Source/WebCore/dom/EventInterfaces.in:
* Source/WebCore/dom/EventNames.json:
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::defaultEventHandler):
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::isValidCommandType):
(WebCore::HTMLDialogElement::handleCommandInternal):
(WebCore::HTMLDialogElement::isValidInvokeAction): Deleted.
(WebCore::HTMLDialogElement::handleInvokeInternal): Deleted.
* Source/WebCore/html/HTMLDialogElement.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::isValidCommandType):
(WebCore::HTMLElement::handleCommandInternal):
(WebCore::HTMLElement::isValidInvokeAction): Deleted.
(WebCore::HTMLElement::handleInvokeInternal): Deleted.
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::commandForElement const):
(WebCore::HTMLFormControlElement::commandType const):
(WebCore::HTMLFormControlElement::handleCommand):
(WebCore::HTMLFormControlElement::invokeTargetElement const): Deleted.
(WebCore::HTMLFormControlElement::invokeAction const): Deleted.
(WebCore::HTMLFormControlElement::handleInvokeAction): Deleted.
* Source/WebCore/html/HTMLFormControlElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::defaultEventHandler):
* Source/WebCore/html/InvokerElement.idl:
* Source/WebCore/page/FragmentDirective.h:

Canonical link: https://commits.webkit.org/281345@main
